core/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java

```diff
@@ -0,0 +1,332 @@
+        }
+        part = envelope.get(partIndex);
+        stepIndex = firstIndex(part.stepCount());
+        position = getStepBeginPos();
+        return true;
+    }
+
+    /** Moves the cursor to the next step */
+    public NextStepResult nextStep() {
+        if (hasReachedEnd())
+            return NEXT_REACHED_END;
+
+        assert partIndex != -1;
+        assert stepIndex != -1;
+        stepIndex = nextIndex(stepIndex, part.stepCount());
+        if (stepIndex == -1)
+            return nextPart() ? NEXT_PART : NEXT_REACHED_END;
+        position = getStepBeginPos();
+        revision++;
+        return NEXT_STEP;
+    }
+
+    public double interpolateSpeed() {
+        return part.interpolateSpeed(stepIndex, position);
+    }
+
+    /**
+     * Scan until the step which contains position
+     * @return whether the step was found
+     */
+    public boolean findPosition(double newPosition) {
+        if (hasReachedEnd())
+            return false;
+
+        // the new position can only move the cursor forward
+        assert comparePos(newPosition, this.position) >= 0;
+
+        // find the EnvelopePart which contains the new position
+        while (comparePos(getPartEndPos(), newPosition) <= 0)
+            if (!nextPart())
+                return false;
+
+        // make sure the new position is within the bounds of the found part
+        assert comparePos(getPartBeginPos(), newPosition) <= 0;
+
+        // now that we found the envelope part, find the exact step
+        while (comparePos(getStepEndPos(), newPosition) <= 0)
+            if (nextStep() == NEXT_REACHED_END)
+                return false;
+
+        // ensure the new position is in the step
+        assert comparePos(newPosition, getStepBeginPos()) >= 0
+                && comparePos(newPosition, getStepEndPos()) <= 0;
+
+        this.position = newPosition;
+        revision++;
+        return true;
+    }
+
+    /** Attempts to find a step which matches the given predicate */
+    public boolean findStep(Envelope.TransitionPredicate predicate) {
+        if (hasReachedEnd())
+            return false;
+
+        do {
+            if (predicate.test(getStepBeginPos(), getStepBeginSpeed(), getStepEndPos(), getStepEndSpeed()))
+                return true;
+        } while (nextStep() != NEXT_REACHED_END);
+        return false;
+    }
+
+    /** Attempts to find a transition between envelope parts which matches a predicate */
+    public boolean findPartTransition(Envelope.TransitionPredicate predicate) {
+        if (hasReachedEnd())
+            return false;
+
+        do {
+            // find the next part
+            var nextPartI = nextPartIndex();
+            if (nextPartI == -1)
+                return false;
+            var nextPart = envelope.get(nextPartI);
+
+            var curEndIndex = lastIndex(part.stepCount());
+            var nextStartIndex = firstIndex(nextPart.stepCount());
+
+            // move to the end of the current part
```

Comment:
"move the cursor to the end of the current part"
To make clear that these two lines are actually modifying the cursor